### ISSUE TITLE
Problems with Servo icon in Windows taskbar

### ISSUE THREAD

**Issue by atouchet** [servo/servo#16326]
When running Servo in Windows 10 the taskbar icon does not always display properly. Sometimes it is displayed properly, sometimes it is displayed with graphical artifacts, and sometimes it does not load at all.

Example cases:

![Windows Servo Icon](https://i.imgur.com/B7UJhA3.png "Windows Servo Icon")

The Servo icon displayed properly.

![Windows Servo Icon](https://i.imgur.com/G6jrCpE.png "Windows Servo Icon")

The Servo icon displayed with graphical artifacts. I have seen two variations so far, one with black splotches in the image (pictured) and another that shows a black square in the middle of the icon image.

![Windows Servo Icon](https://i.imgur.com/krvYmgI.png "Windows Servo Icon")

The Servo icon not being loaded at all.

**Comment by atouchet** [comment]
This may be improved/fixed by #11969.

**Comment by Coder206** [comment]
@atouchet I have been able to reproduce this (as #11969 shows). However, it only appears for me when I run Servo with `mach.bat build [-d|-r]`. When I run it from the target directory, it looks fine.

**Comment by jonathandturner** [comment]
Huh, do we know why the release one is having problems?

Also what are the steps to repro?  I'll be putting the nightly through it's paces once the last two issues are resolved to see if it's ready to go live very soon, maybe as soon as tomorrow.

**Comment by atouchet** [comment]
I haven't found a consistent way to reproduce this issue yet. I can load the exact same build and get different results at different times.

**Comment by codec-abc** [comment]
In my experience there is 2 ways to set up an icon in Windows. The first is to set it a build time. I am no expert but there I think it end in the IMAGE_RESOURCE_DIRECTORY_ENTRY region of the Portable Executable (PE) header of the exe file. The second method is to set it dynamically at run-time using the Windows API. Contrary to Firefox, it seems that Servo use the 2nd method. I can tell that because when looking in the explorer the exe file of Firefox has an icon while Servo has the default executable icon. From my experience in C# doing Winform app, the 1st method give cleaner result. I did the 2nd method once and while I had always the icon, there was some pixel aliasing issue. So, I guess there is some mip-mapping in the PE header allowing windows to display it nicely and when given an icon at run-time Windows would scale it on fly with poor result.

TLDR: It would be better if the icon was embedded in the icon at build time instead of being loaded at runtime.

**Comment by atouchet** [comment]
I should maybe also note that the issue in the original post is in regards to the installer version of Servo. The standalone ZIP version does not load the icon at all.

**Comment by atouchet** [comment]
This was fixed in #20316.
